src/t_stream.c

```diff
@@ -632,6 +721,321 @@ int64_t streamTrimByLength(stream *s, size_t maxlen, int approx) {
+    }
+    raxStop(&ri);
+
+    return deleted;
+}
+
+#define TRIM_STRATEGY_NONE 0
+#define TRIM_STRATEGY_MAXLEN 1
+#define TRIM_STRATEGY_MINID 2
+
+typedef struct {
+    /* XADD options */
+    streamID id; /* User-provided ID, for XADD only. */
+    int id_given; /* Was an ID different than "*" specified? for XADD only. */
+
+    /* XADD + XTRIM common options */
+    int trim_strategy; /* TRIM_STRATEGY_* */
+    int approx; /* If 1 only delete whole radix tree nodes, so
+                 * the trim argument is not applied verbatim. */
+    long long maxnodes; /* Maximum amount of node to trim. If 0 no limitation
+                         * on the amount of trimming work is enforced. */
+    /* TRIM_STRATEGY_MAXLEN options */
+    long long maxlen; /* After trimming, leave stream at this length . */
+    int maxlen_arg_idx; /* Index of the count in MAXLEN, for rewriting. */
+    /* TRIM_STRATEGY_MAXLEN options */
+    streamID minid; /* Trim by ID (No stream entries with ID < 'minid' will remain) */
+    int minid_arg_idx; /* Index of the count in MINID, for rewriting. */
+    int no_mkstream; /* if set to 1 do not create new stream */
+} streamAddTrimArgs;
+
+static int streamParseAddOrTrimArguments(client *c, streamAddTrimArgs *args, int xadd) {
+    /* Initialize arguments to defaults */
+    memset(args, 0, sizeof(*args));
+
+    /* Parse options. */
+    int i = 2; /* This is the first argument position where we could
+                  find an option, or the ID. */
+    int maxnodes_given = 0;
+    for (; i < c->argc; i++) {
+        int moreargs = (c->argc-1) - i; /* Number of additional arguments. */
+        char *opt = c->argv[i]->ptr;
+        if (xadd && opt[0] == '*' && opt[1] == '\0') {
+            /* This is just a fast path for the common case of auto-ID
+             * creation. */
+            break;
+        } else if (!strcasecmp(opt,"maxlen") && moreargs) {
+            args->approx = 0;
+            char *next = c->argv[i+1]->ptr;
+            /* Check for the form MAXLEN ~ <count>. */
+            if (moreargs >= 2 && next[0] == '~' && next[1] == '\0') {
+                args->approx = 1;
+                i++;
+            } else if (moreargs >= 2 && next[0] == '=' && next[1] == '\0') {
+                i++;
+            }
+            if (getLongLongFromObjectOrReply(c,c->argv[i+1],&args->maxlen,NULL)
+                != C_OK) return -1;
+
+            if (args->maxlen < 0) {
+                addReplyError(c,"The MAXLEN argument must be >= 0.");
+                return -1;
+            }
+            i++;
+            args->trim_strategy = TRIM_STRATEGY_MAXLEN;
+            args->maxlen_arg_idx = i;
+        } else if ((!strcasecmp(opt,"minid") || !strcasecmp(opt,"maxage")) && moreargs) {
+            args->approx = 0;
+            char *next = c->argv[i+1]->ptr;
+            /* Check for the form MINID|MAXAGE ~ <id>|<age>. */
+            if (moreargs >= 2 && next[0] == '~' && next[1] == '\0') {
```

Comment:
the problem is that i don't know where exactly the trim_strategy will be, because i could get NOMKSTREAM for example of `*`
if this function would handle only trim strategies (i.e. the first arg is always the trim strategy) then it would make more sense to parse like you suggested

Reply:
I think you're right. Refactoring like this won't make it simpler.